arch/arm/src/stm32wl5/stm32wl5_spi.c

```diff
@@ -0,0 +1,1859 @@
+  void *xbuffer = rxbuffer;
+  int ret;
+
+  DEBUGASSERT(priv != NULL);
+
+  /* Convert the number of word to a number of bytes */
+
+  size_t nbytes = (priv->nbits > 8) ? nwords << 1 : nwords;
+
+#ifdef CONFIG_STM32WL5_SPI_DMATHRESHOLD
+  /* If this is a small SPI transfer, then let spi_exchange_nodma() do the
+   * work.
+   */
+
+  if (nbytes <= CONFIG_STM32WL5_SPI_DMATHRESHOLD)
+    {
+      spi_exchange_nodma(dev, txbuffer, rxbuffer, nwords);
+      return;
+    }
+#endif
+
+  if ((priv->rxdma == NULL) || (priv->txdma == NULL) ||
+      up_interrupt_context())
+    {
+      /* Invalid DMA channels, or interrupt context, fall
+       * back to non-DMA method.
+       */
+
+      spi_exchange_nodma(dev, txbuffer, rxbuffer, nwords);
+      return;
+    }
+
+#ifdef CONFIG_STM32WL5_DMACAPABLE
+  if ((txbuffer && priv->txbuf == 0 &&
+      !stm32wl5_dmacapable((uintptr_t)txbuffer, nwords, priv->txccr)) ||
+      (rxbuffer && priv->rxbuf == 0 &&
+       !stm32wl5_dmacapable((uintptr_t)rxbuffer, nwords, priv->rxccr)))
+    {
+      /* Unsupported memory region fall back to non-DMA method. */
+
+      spi_exchange_nodma(dev, txbuffer, rxbuffer, nwords);
+    }
+  else
+#endif
+    {
+      static uint16_t rxdummy = 0xffff;
+      static const uint16_t txdummy = 0xffff;
+
+      spiinfo("txbuffer=%p rxbuffer=%p nwords=%d\n",
+               txbuffer, rxbuffer, nwords);
+      DEBUGASSERT(priv && priv->spibase);
+
+      /* Setup DMAs */
+
+      /* If this bus uses a in driver buffers we will incur 2 copies,
+       * The copy cost is << less the non DMA transfer time and having
+       * the buffer in the driver ensures DMA can be used. This is because
+       * the API does not support passing the buffer extent so the only
+       * extent is buffer + the transfer size. These can sizes be less than
+       * the cache line size, and not aligned and typically greater then 4
+       * bytes, which is about the break even point for the DMA IO overhead.
+       */
+
+      if (txbuffer && priv->txbuf)
+        {
+          if (nbytes > priv->buflen)
+            {
+              nbytes = priv->buflen;
+            }
+
+          memcpy(priv->txbuf, txbuffer, nbytes);
+          txbuffer = priv->txbuf;
+          rxbuffer = !rxbuffer ? priv->rxbuf : rxbuffer;
+        }
+
+      spi_dmarxsetup(priv, rxbuffer, &rxdummy, nwords);
+      spi_dmatxsetup(priv, txbuffer, &txdummy, nwords);
+
+#ifdef CONFIG_SPI_TRIGGER
+      /* Is deferred triggering in effect? */
+
+      if (!priv->defertrig)
+        {
+          /* No.. Start the DMAs */
+
+          spi_dmarxstart(priv);
+          spi_dmatxstart(priv);
+        }
+      else
+        {
+          /* Yes.. indicated that we are ready to be started */
+
+          priv->trigarmed = true;
+        }
+#else
+      /* Start the DMAs */
+
+      spi_dmarxstart(priv);
+      spi_dmatxstart(priv);
+#endif
+
+      /* Then wait for each to complete */
+
+      ret = spi_dmarxwait(priv);
+      if (ret < 0)
```

Comment:
The `spi_dmarxwait` implementation needs to be taken into account. See comments there.

Reply:
I'm not sure if need to check `ret` here at all. I think that the best way is just to
```
spi_dmarxwait(priv);
spi_dmatxwait(priv);
```
The return value is always `OK` from both of the APIs, so we can make those `void`.
We do not have the way to cancel DMA transaction in case of error, so we just wait for successful completion always

Reply:
@onegray  I'm not sure about Your suggestion. If the **ret >= 0** then all is OK and we can make memcpy. In if statemant we are trying to retry operation in order to get **ret** value that is >= 0.

Reply:
I think it does not matter since wait function can only return OK

Reply:
Since we have a return code, then checking it accordingly also looks good for me.